### ISSUE TITLE
Improve error message when there's no space before semicolon (PEP 508 URL requirement)

### ISSUE THREAD

**Issue by hauntsaninja** [pypa/packaging#494]
I'm not fully sure whether this is packaging's fault (or importlib.metadata or pip or setuptools), but here's what I'm seeing:

```
mkdir reqparse
cd reqparse

cat > setup.py <<HERE
from setuptools import setup

setup(
    name="reqparse",
    extras_require={
        "test": [
            "gym-retro @ git+https://github.com/openai/retro.git@v0.8.0",
        ]
    },
)
HERE

python3 -m venv env
source env/bin/activate
python3 -m pip install packaging -e .

cat reqparse.egg-info/requires.txt

# this now gives me
# packaging.requirements.InvalidRequirement: Parse error at "'extra =='": Expected string_end
python3 -c '
import importlib.metadata
from packaging.requirements import Requirement
Requirement(importlib.metadata.distribution("reqparse").requires[0])
'
```

Or more concisely, but with less justification:
```
from packaging.requirements import Requirement
Requirement('gym-retro@ git+https://github.com/openai/retro.git@v0.8.0; extra == "test"')
```

Thank you!

**Comment by uranusjr** [comment]
It’s likely importlib.metadata’s fault. It automatically converts egg-info-style requirements (which is used by legacy editables) to PEP 508-style, and the conversion went wrong.

cc @jaraco the easiest fix would be to add a space before the semicolon here

https://github.com/python/cpython/blob/908fd691f96403a3c30d85c17dd74ed1f26a60fd/Lib/importlib/metadata/__init__.py#L678

**Comment by brettcannon** [comment]
I can't transfer across orgs, but the repo is https://github.com/python/importlib_metadata .

**Comment by hauntsaninja** [comment]
Thanks both! :-) Created a new issue in importlib_metadata

**Comment by jaraco** [comment]
https://github.com/python/importlib_metadata/issues/357

**Comment by pradyunsg** [comment]
I'm gonna reopen this, to track adding a better error message for this; especially once #484 is merged.

**Comment by pradyunsg** [comment]
Closing this, since there's a more targetted issue for this in #529 now.
